### PR TITLE
feat(agent): Keras Documentation reference agent

### PR DESCRIPTION
**Pull request by soumik12345**
This PR implements a `KerasDocumentationAgent` agent that would use the `KerasIORetreiver` to retrieve respective documentation sources in a structured format according to a schema.

[Sample vector index artifact](https://wandb.ai/ml-colabs/ml-frameworks-bot/artifacts/vector_index/keras3_api_reference/v4/overview) | [Sample retrieval trace](https://wandb.ai/ml-colabs/ml-frameworks-bot/weave/calls/0191f1ab-b927-7062-90cc-9ae3b70b756e)

### CODE REVIEW

File: ml_frameworks_bot/keras_io/docs_agent.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                    "role": "system",
+                    "content": """
+You are an experienced machine learning engineer expert in python and Keras.
+You are suppossed to think step-by-step about all the unique Keras operations,
+layers, and functions from a given snippet of code.
+
+Here are some rules:
+1. All functions and classes that are imported from `keras` should be considered to
+    be Keras operations.
+2. `impport` statements don't count as separate statements.
```
Comment (SauravMaheshkar) [inline ml_frameworks_bot/keras_io/docs_agent.py]:
```suggestion
2. `import` statements don't count as separate statements.
```

Minor typo

File: ml_frameworks_bot/keras_io/docs_agent.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            ops_with_api_reference.append(
+                KerasOpWithAPIReference(
+                    keras_op=keras_op, api_reference=api_reference.text
+                )
+            )
+        return ops_with_api_reference
+
+    @weave.op()
+    def predict(
+        self, code_snippet: str, seed: Optional[int] = None, max_retries: int = 3
```
Comment (SauravMaheshkar) [inline ml_frameworks_bot/keras_io/docs_agent.py:90]: Minor: maybe we could make `seed` and `max_retires` as parameters when instantiating the class.

File: ml_frameworks_bot/keras_io/retriever.py [hunk 3]
```diff
@@ -72,27 +85,42 @@ def load_documents(self, num_workers: Optional[int] = None) -> List[Document]:
...
-            included_file_extensions=[".md"],
-        )
-        reader = SimpleDirectoryReader(input_files=input_files)
-        return reader.load_data(num_workers=num_workers, show_progress=True)
+        source_directory = os.path.join(self.repository_local_path, "sources")
+        if not os.path.exists(source_directory):
+            build_keras_io_sources(repository_local_path=self.repository_local_path)
+        input_files = []
+        for directory in included_directories:
+            input_files += get_all_file_paths(
```
Comment (SauravMaheshkar) [inline ml_frameworks_bot/keras_io/retriever.py:105]: could we do this with [glob](https://docs.python.org/3/library/glob.html) instead ?

File: ml_frameworks_bot/keras_io/retriever.py [hunk 4]
```diff
@@ -56,13 +62,20 @@ def from_wandb_artifact(cls, artifact_address: str, similarity_top_k: int = 10):
...
         )
+        _cls.wandb_artifact_address = artifact_address
+        return _cls
 
-    def load_documents(self, num_workers: Optional[int] = None) -> List[Document]:
+    def load_documents(
+        self,
+        included_directories: List[str] = ["examples", "guides", "templates"],
+        exclude_file_postfixes: List[str] = ["index.md"],
+        return_nodes: bool = True,
```
Comment (SauravMaheshkar) [inline ml_frameworks_bot/keras_io/retriever.py:89]: Question: why the `return_nodes` parameter? Shouldn't we just do text nodes for a single reference?

File: ml_frameworks_bot/keras_io/retriever.py [hunk 5]
```diff
@@ -102,43 +130,76 @@ def chunk_documents(
...
                 )
-                if build_index_from_documents
-                else VectorStoreIndex(nodes=nodes, show_progress=True)
+                if apply_chunking
+                else document_nodes
             )
+            self._vector_index = VectorStoreIndex(nodes=document_nodes)
+            print(f"{len(document_nodes)=}")
+            print(f"{len(self._vector_index.docstore.docs)=}")
+            assert len(document_nodes) == len(self._vector_index.docstore.docs)
```
Comment (SauravMaheshkar) [inline ml_frameworks_bot/keras_io/retriever.py]:
```suggestion
            assert len(document_nodes) == len(self._vector_index.docstore.docs), f"No. of document nodes {len(document_nodes)} != No. of nodes in VectorIndex {len(self._vector_index.docstore.docs)}"
```

File: ml_frameworks_bot/utils.py [hunk 6]
```diff
@@ -51,5 +54,28 @@ def make_embedding_model(embedding_model_name: str) -> BaseEmbedding:
             "text-embedding-3-large",
             "text-embedding-ada-002",
         ]
-        else HuggingFaceEmbedding(model_name=embedding_model_name)
+        else HuggingFaceEmbedding(
+            model_name=embedding_model_name,
+            trust_remote_code=True,
+            model_kwargs={"torch_dtype": torch.float16},
```
Comment (SauravMaheshkar) [inline ml_frameworks_bot/utils.py]: `model_kwargs` should be a optional parameter for the function. 

Example usage:
```python
embedding_model = make_embedding_model("text-embedding-3-small", **hf_kwargs)
```

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+      - requirements.txt
+      - setup.py
+
+jobs:
+  code-format:
+    name: check code format using black
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: psf/black@stable
```
Comment (SauravMaheshkar) [inline .github/workflows/tests.yml:16]: Comment: We should migrate to `ruff format` later so that we're just using a single tool.

File: README.md [hunk 8]
```diff
@@ -2,42 +2,84 @@
 
 An agentic workflow to translate machine learning codebases across ML frameworks reliably at scale.
 
-## Build Vector Index for [keras.io](https://keras.io/)
-
+<details>
+<summary>Build Vector Index for <a href="https://keras.io/">keras.io</a></summary>
```
Comment (SauravMaheshkar) [inline README.md:6]: Nice ❤️

**Review by SauravMaheshkar — CHANGES_REQUESTED** [review]

File: ml_frameworks_bot/keras_io/docs_agent.py (hunk 2 above)
Comment (soumik12345) [reply ml_frameworks_bot/keras_io/docs_agent.py:90]: I disagree with this suggestion. The model once initialized in production stays initialized with the same state (until the prod env is rebooted). Attributes like `llm_name` are part of the state because that is ideally not gonna be mutable in prod (we're gonna go with the llm vendor that works the best as per our evaluations), whereas `seed` and `max_retries` should be made to be tuned as per the query.

File: ml_frameworks_bot/keras_io/retriever.py (hunk 3 above)
Comment (soumik12345) [reply ml_frameworks_bot/keras_io/retriever.py:105]: I don't think `glob` recursively fetches all files in a directory. Given that in our case, the depth of the directory tree is variable, `glob` can be used.

File: .github/workflows/tests.yml (hunk 7 above)
Comment (soumik12345) [reply .github/workflows/tests.yml:16]: Good catch. You can ship this with the PyTorch retriever PR.

File: ml_frameworks_bot/keras_io/docs_agent.py (hunk 2 above)
Comment (SauravMaheshkar) [reply ml_frameworks_bot/keras_io/docs_agent.py:90]: Makes sense.

File: ml_frameworks_bot/keras_io/retriever.py (hunk 4 above)
Comment (soumik12345) [reply ml_frameworks_bot/keras_io/retriever.py:89]: I left an optional flag `apply_chunking` in `KerasIORetreiver.index_documents`, which optionally applies semantic chunking to the documents before generating the embeddings. When `apply_chunking=True`, we need `Document`s instead of `TextNode`s.

**Review by SauravMaheshkar — APPROVED** [review]
LGTM !
